streamer/m3u8_concater.py

```diff
@@ -0,0 +1,957 @@
+          self.content += line
+          line = media_playlist.readline()
+          # If a byterange exists, add it to the content.
+          if line.startswith('#EXT-X-BYTERANGE'):
+            self.content += line
+            line = media_playlist.readline()
+          # Update the segment's URI.
+          self.content += os.path.join(period_dir ,line)
+        elif line.startswith('#EXT-X-MAP'):
+          # An EXT-X-MAP must have a URI attribute and optionally
+          # a BYTERANGE attribute.
+          attribs = _extract_attributes(line)
+          self.content += '#EXT-X-MAP:URI=' + _quote(
+              os.path.join(period_dir, _unquote(attribs['URI'])))
+          if attribs.get('BYTERANGE'):
+            self.content += ',BYTERANGE=' + attribs['BYTERANGE']
+          self.content += '\n'
+        elif line.startswith(MediaPlaylist.HEADER_TAGS + ('#EXT-X-ENDLIST',)):
+          # Escape header and end-list tags.
+          pass
+        elif not line.startswith('#EXT'):
+          # Escape comments.
+          pass
+        elif line.startswith('#EXT-X-TARGETDURATION'):
+          self.target_duration = int(line[len('#EXT-X-TARGETDURATION:'):])
+        else:
+          # Store lines that didn't match one of the above cases.
+          # Like ENCRYPTIONKEYS, DISCONTINUITIES, COMMENTS, etc... .
+          self.content += line
+        line = media_playlist.readline()
+    
+    # Set the features we need to access easily while performing the concatenation.
+    # Features like codec, channel_layout, resolution, etc... .
+    self._set_features(streams_map)
+  
+  def _set_features(self, streams_map: Dict[str, OutputStream]) -> None:
+    """Get the audio and video codecs and other relavent stream features
+    from the matching OutputStream in the `streams_map`, this will be used
+    in the codec matching process in the concat_xxx() methods, but the codecs
+    that will be written in the final concatenateded master playlist will be
+    the codecs from stream_info dictionary(in HLS syntax).
+    """
+    
+    # We can't depend on the stream_info['CODECS'] to get the codecs from because
+    # this is only present for STREAM-INF, this makes it harder to get codecs for
+    # audio segments. Also if we try to get the audio codecs from one of the 
+    # #EXT-X-STREAM-INF tags we will have to match these codecs with each stream,
+    # for example, a codec attribute might be CODECS="avc1,ac-3,mp4a,opus",
+    # the video codec is put first then the audio codecs are put in 
+    # lexicographical order(by observation), which isn't nesseccary the same order of 
+    # #EXT-X-MEDIA in the master playlist, thus there is no solid baseground for 
+    # matching the codecs using the information in the master playlist.
+    
+    output_stream: Optional[OutputStream] = None
+    lines = self.content.split('\n')
+    for i in range(len(lines)):
+      # Don't use #EXT-X-MAP to get the codec, because HLS specs says
```

Comment:
```python
def get_init_seg_file(self) -> Pipe:
    INIT_SEGMENT = {
      MediaType.AUDIO: 'audio_{language}_{channels}c_{bitrate}_{codec}_init.{format}',
      MediaType.VIDEO: 'video_{resolution_name}_{bitrate}_{codec}_init.{format}',
      MediaType.TEXT: 'text_{language}_init.{format}',
    }
    path_templ = INIT_SEGMENT[self.type].format(**self.features)
    return Pipe.create_file_pipe(path_templ, mode='w')

  def get_media_seg_file(self) -> Pipe:
    MEDIA_SEGMENT = {
      MediaType.AUDIO: 'audio_{language}_{channels}c_{bitrate}_{codec}_$Number$.{format}',
      MediaType.VIDEO: 'video_{resolution_name}_{bitrate}_{codec}_$Number$.{format}',
      MediaType.TEXT: 'text_{language}_$Number$.{format}',
    }
    path_templ = MEDIA_SEGMENT[self.type].format(**self.features)
    return Pipe.create_file_pipe(path_templ, mode='w')

  def get_single_seg_file(self) -> Pipe:
    SINGLE_SEGMENT = {
      MediaType.AUDIO: 'audio_{language}_{channels}c_{bitrate}_{codec}.{format}',
      MediaType.VIDEO: 'video_{resolution_name}_{bitrate}_{codec}.{format}',
      MediaType.TEXT: 'text_{language}.{format}',
    }
    path_templ = SINGLE_SEGMENT[self.type].format(**self.features)
    return Pipe.create_file_pipe(path_templ, mode='w')
```
```m3u8
#EXTM3U
#EXT-X-VERSION:6
## Generated with https://github.com/google/shaka-packager version c1f64e5-release
#EXT-X-TARGETDURATION:4
#EXT-X-PLAYLIST-TYPE:VOD
#EXT-X-MAP:URI="video_480p_1M_h264_init.mp4"
#EXTINF:3.000,
video_480p_1M_h264_1.mp4
#EXT-X-ENDLIST
```
We use the `streams_map` to get the matching output stream object, but when creating the streams map we only added the mapping from the methods `get_single_seg_file()` and `get_media_seg_file()`, in the `segment_per_file` case, we could have added `get_init_seg_file()` instead and checked the URI in the `#EXT-X-MAP` tag, this would yield to the same result, but i was skeptic about that in some cases packager might not output an `#EXT-X-MAP` tag since it's optional according to the HLS specs.  Kept the comment as a reminder.

Reply:
Ah, now I see.  How about this, which would make more sense to me in a comment:

```
# Don't use the URIs from any tag to try to extract codec information.
# We should not rely on the exact structure of file names for this.
# Use stream_maps instead.
```